materialize-sql/type_mapping.go

```diff
@@ -193,6 +193,33 @@ func WithElementConverter(converter ElementConverter) StaticMapperOption {
 	}
 }
 
+// FieldConfigWrapper wraps a TypeMapper to specify the type of the column in
+// the case that it is a primary key. This is useful for cases where specific
+// databases require certain variations of a type for primary keys. For example,
+// MySQL does not accept TEXT as a primary key, but a VARCHAR with specified
+// size works.
+type FieldConfigMapper struct {
+	Delegate TypeMapper
+	Map      func(*MappedType, json.RawMessage) error
+}
+
+var _ TypeMapper = FieldConfigMapper{}
+
+func (m FieldConfigMapper) MapType(p *Projection) (mapped MappedType, err error) {
+	delegated, err := m.Delegate.MapType(p)
+	if err != nil {
+		return MappedType{}, err
+	}
+
+	if p.RawFieldConfig != nil {
```

Comment:
nit: an improvement would be to have `FieldConfigMapper` have an argument function which returns a new instance of the field config to parse into.

Then the deserialization and error check logic can live here, and the `Map` function can receive a pre-validated instance of it's custom field config type.

Reply:
Updated it to use a generic type which allows it to handle the parsing of the config and pass it to the map function (without a need to later do type assertions on an `interface`

Reply:
This is even better, I keep forgetting that Go generics are now a thing :+1: